### PR TITLE
Add segmentation map processing to SAM Image Processor

### LINKED ISSUE #27361: Add how to preprocess mask for finetuning with SAM
### Feature request

The [SAM image processor](https://github.com/huggingface/transformers/blob/main/src/transformers/models/sam/image_processing_sam.py) takes images as input and resizes them so that the longest edge is 1024 (using default values). This is the size expect as input fo the SAM model. 
For inference, this works fine as only the images need resizing but for fine-tuning as per [this tutorial](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb), you need to resize both your images and your masks as the SAM model produces `pred_masks` with size 256x256. If I don't resize my masks I get `ground truth has different shape (torch.Size([2, 1, 768, 1024])) from input (torch.Size([2, 1, 256, 256]))` when trying to calculate loss.

To fix this, I've currently written a resize and pad function into my code:

```
from PIL import Image

def resize_mask(image):
    longest_edge = 256
    
    # get new size
    w, h = image.size
    scale = longest_edge * 1.0 / max(h, w)
    new_h, new_w = h * scale, w * scale
    new_h = int(new_h + 0.5)
    new_w = int(new_w + 0.5)

    resized_image = image.resize((new_w, new_h), resample=Image.Resampling.BILINEAR)
    return resized_image

def pad_mask(image):
    pad_height = 256 - image.height
    pad_width = 256 - image.width

    padding = ((0, pad_height), (0, pad_width))
    padded_image = np.pad(image, padding, mode="constant")
    return padded_image

def process_mask(image):
    resized_mask = resize_mask(image)
    padded_mask = pad_mask(resized_mask)
    return padded_mask
```

and then have added this to my definition of SAMDataset:

```
class SAMDataset(Dataset):
    def __init__(self, dataset, processor, transform = None):
        self.dataset = dataset
        self.processor = processor
        self.transform = transform

    def __len__(self):
        return len(self.dataset)

    def __getitem__(self, idx):
        item = self.dataset[idx]
        
        if self.transform:
            image = self.transform(item["pixel_values"])
        else:
            image = item["pixel_values"]
        
        # get bounding box prompt
        padded_mask = process_mask(item["label"])
        prompt = get_bounding_box(padded_mask)

        # prepare image and prompt for the model
        inputs = self.processor(image, input_boxes=[[prompt]], return_tensors="pt")

        # remove batch dimension which the processor adds by default
        inputs = {k:v.squeeze(0) for k,v in inputs.items()}

        # add ground truth segmentation
        inputs["ground_truth_mask"] = padded_mask

        return inputs
```

This seems to work fine. 

What I think would be good is to allow input of masks in the SAM image processor. For example, the [Segformer image processor](https://github.com/huggingface/transformers/blob/v4.35.0/src/transformers/models/segformer/image_processing_segformer.py#L305) takes images and masks as inputs and resizes both to the size expected by the Segformer model. 

I have also seen there is a 'post_process_mask' method in the SAM image processor but I am unsure how to implement this in the tutorial I'm following. If you think this is a better way vs. what I am suggesting then please could you explain where I would add this in the code from the tutorial notebook.

### Motivation

Easier fine tuning of SAM model.

### Your contribution

I could try write a PR for this and/or make a PR to update the [notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb) instead .

### PR DESCRIPTION
**Pull request by rwood-97**
# What does this PR do?

Fixes #27361 

## Before submitting
- [ ] ~This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).~
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts

### CODE REVIEW

**Comment by amyeroberts** [comment]
@rwood-97 - thanks for opening this PR - let me know when it's ready for review!

**Comment by rwood-97** [comment]
@amyeroberts I've added a very small amount to the tests. I think they could be more comprehensive but I think at min this now tests a case of input with/without masks. Let me know if there are any others you'd want me to add.

**Comment by rwood-97** [comment]
Hey, just looking through the current documentation on SAM and I think the most relevant place for me to add docs is probably [this tutorial](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb). I'll create an issue on that repo once this PR is done so I think this PR could be considered as ready now.

**Review by amyeroberts — APPROVED** [review]
Beautiful PR 🤩 

Thanks for adding - especially for sticking so closely with other implementation patterns and adding tests too ❤️ !

**Comment by amyeroberts** [comment]
@rwood-97 Regarding where to document, the best place would be to add an example on the [model doc page](https://github.com/huggingface/transformers/blob/main/docs/source/en/model_doc/sam.md) and/or in the officially supported notebooks - [here](https://github.com/huggingface/notebooks/blob/main/examples/segment_anything.ipynb) and [here](https://github.com/huggingface/notebooks/blob/main/examples/automatic_mask_generation.ipynb). 

Could you add a small code snippet to the model doc page to show how to pass and use the masks? After that we can merge :) 

The notebook you linked to isn't officially maintained by Hugging Face. You can certainly still open a PR to add to it if you'd like!

**Comment by rwood-97** [comment]
yep sure, I'll try find some time this week/next week for this

File: docs/source/en/model_doc/sam.md
Comment (rwood-97) [inline docs/source/en/model_doc/sam.md:1]: Can't seem to comment on the exact line but should line 36 be updated to say this model does support fine-tuning?

Comment (amyeroberts) [reply docs/source/en/model_doc/sam.md:1]: Yes, it's annoying that github won't let you comment much outside the diff.

In answer to your question, not yet. There's a series of checks we'd have to do e.g. enabling test_training which are outside the scope of this PR. If you'd like you're welcome to tackle it in a separate PR or open an issue as a feature request

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating! 

Just one final comment

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating on this - looking great! 

Just one last thing before we can merge. Could you add segmentation_maps as an explicit argument to the SAM processor's `__call__` method? At the moment this works  because all the kwargs are forwarded to the image processor, but we want to make this more explicit and reduce the use of kwargs where possible.

**Comment by rwood-97** [comment]
All done 👍

**Comment by amyeroberts** [comment]
@rwood-97 Thanks again for all the work on this contribution!
